### DISCUSSION TITLE
In Svelte 5, you could make custom signals that also work as Observables at the same time

### DISCUSSION THREAD

**Discussion by Evertt** [state-adapt/state-adapt#97] [General]
So I was just playing around again with Svelte 5 and creating custom signals built upon Svelte 5 runes and I realized that you can actually make a signal that can both function as a signal and as a store, at the same time.

Which at first glance, seems really great to me. Because of course, when you're creating custom signals in Svelte 5, you don't get the sweet ergonomics that the compiler gives to runes and you'll need to either use .value or call your signal as a function (like in SolidJS) for your own custom signals. But if you give your own custom signal a .subscribe() method too, then you can just read the value of your signal in your code using the $ prefix. Which looks much more pleasing to my eyes.

And on top of that, if you actually base those custom signals / stores on RxJS observables, then suddenly you can take advantage of all the advance features of RxJS (i.e. mainly being able to pipe your signal through a bunch of operators.

## Edit

[Here's the original demo](https://svelte.dev/playground/b6b4d03d258f4cb2a0248b0859ae3f28?version=5.16.5)

[Here's what I'm working on right now](https://svelte.dev/playground/ba36e0f04f4b4ea1963ccc7731cc1f14?version=5.16.0)

I'm curious if you like it as much as I do. 😁

**Comment by mfp22** [comment]
Awesome. That's really versatile. 

> Which looks much more pleasing to my eyes.

It could also make it so refactoring from stores to custom signals wouldn't require touching the template, right?

And is there a way to wrap `$derived` in something that would let you consume the result like a store too?

**Comment by Evertt** [comment]
> It could also make it so refactoring from stores to custom signals wouldn't require touching the template, right?

Correct. Only the plain `count` syntax is a bit fragile, because when it's used in the template by itself (without any other reactive variable nearby) then the Svelte 5 compiler will assume that `count` isn't reactive and it won't put it into a `render_effect()` statement.

> And is there a way to wrap `$derived` in something that would let you consume the result like a store too?

Don't know, would need to experiment with that, but I'm not sure when I'll be able to do that. At the moment I'm in a pretty stressful situation where I've been out of work for 1½ years already, desperately looking for a job, and I have a long outstanding debt with the tax office of over 12,000 euros. So for now I need to focus all my attention on finding an income asap.

**Reply by mfp22** [reply]
Hey, just checking in. Are you doing okay now? I just finished up a project and will have a lot of time for StateAdapt for the next few months. But did you find a project? A lot of devs I know are looking for work as well. Rough time right now

**Reply by Evertt** [reply]
Unfortunately no, things have gotten much worse actually. I'm pretty suicidal at the moment to be frank. I'm getting some help from mental health professionals, but for now I don't have much hope. They have a plan for me to get better and I'm gonna try it and hope that I'll make it to the end and that it will actually help.

### Edit

Oh yeah, I haven't been able to really find a job, let alone keep it. Because the first 2 weeks I am super productive, I tend to deliver approximately the same amount of work that other people need 5 weeks for. But after those 2 weeks, my performance drops significantly and then the people I work for are so disappointed that I get fired pretty quickly.

**Reply by Evertt** [reply]
Oh by the way, I was trying to improve my code a bit. So that it would truly return to an unresolved promise when unmounted, so that the `{#await}` block renders correctly. But I seem to have broken my code and I really don't understand how. 😅

**Reply by mfp22** [reply]
I'm really sorry to hear that. I hope you're able to get through this and find meaning. I'm not good at helping with this stuff, but I can only say what has helped me in the past. A lot of people don't like Jordan Peterson, but I found his original Maps of Meaning lectures on YouTube to be really powerful. It's devoid of controversial politics, and there's a lot of psychology knowledge to learn from them. It helped me know what to look for for a more meaningful life.

**Comment by Evertt** [comment]
New update, I've been tinkering with my code again, to try to get it to work as expected again, and well, I think I managed it.

https://svelte.dev/playground/ba36e0f04f4b4ea1963ccc7731cc1f14?version=5.16.1

**Reply by Evertt** [reply]
Okay, it's still buggy. @mfp22 what do you think of my discussion that I've placed here?

https://github.com/ReactiveX/rxjs/discussions/7524

**Reply by mfp22** [reply]
I see someone responded. I'm not very familiar with RxJS source code, actually. Does his response make sense?

**Reply by Evertt** [reply]
Honestly, I have no idea. I don't understand it well enough.

**Comment by Evertt** [comment]
Next update, I've broken the code again and I have no clue how to fix it. Now, in the newer version of my demo, the `$count` seems to be the only correct value and all the other ones seem to be one step behind. @mfp22 can you figure out why that might be happening?

I have been able to find my original demo, but there I didn't test with piping either, so that might have the same bug too.

**Reply by mfp22** [reply]
Your example is pretty complicated. Did you change the RxJS operator code when you pasted it in? I assume you did when you were thinking there was a bug with the operator code. 

If the example can't be simplified, I'll have to dive deeper tomorrow.

**Reply by Evertt** [reply]
Yeah I changed it to what I thought it should be like. If you change it back then you'll just see different, but still buggy, behavior

**Reply by Evertt** [reply]
And yes I'm making my example complicated on purpose, because I want it to work for every (edge) case.

**Comment by Evertt** [comment]
Anyhow, if you start working on your own `SignalStore` class, you may want to check out [this page from the docs about `createSubscriber()`](svelte.dev/docs/svelte/svelte-reactivity#createsubscriber). Because that could be used to subscribe to the internally held Observable when your class' `value` is being used in a component template or effect.

**Reply by mfp22** [reply]
That makes sense.

StateAdapt seems to be fine for the time being. I need to make a lot of improvements to it. But I'm primarily concerned with RxJS falling away from popularity in Angular codebases right now. Probably 80% of StateAdapt users are from Angular, and in order to grow it and help its long-term health, I need to grapple with the communication problem. So I'm working on an article about reactivity in Angular. I'll have more for all frameworks going forward. But this is a key moment in time.

I'm also mulling over some thoughts about AI. Right now every non-React coding pattern is at a disadvantage. The lowest common denominator might get cemented in for a few years, and that would suck. I think declarative code will help both people and AI. I can set up some kind of experiment perhaps... Context window doesn't need to be as long for declarative code.
